### PR TITLE
Adding enhancements for Disaster Recovery with the introduction of CEO

### PR DESCRIPTION
**Pull request by retroflexer**
Enhancement to the current set of disaster recovery scripts to make the automated tests pass for 4.4.

### CODE REVIEW

**Comment by deads2k** [comment]
These are good thoughts, focused on one of our existing test cases.  I think it's worth updating the `Test Plan` section of our exist enhancement here https://github.com/openshift/enhancements/blame/master/enhancements/etcd/cluster-etcd-operator.md#L475 with high level goals.  A starter list for instance.


1. If one master is lost, instructions for how to.
    1. create a new master that joins the cluster
    2. removal of the old master from the cluster
2. Changes to the etcd-quorum recovery steps
3. All nodes being shut off at the same time and restarted.
4. IP address change of a single member
5. IP address change of all members
6. debugging and detection when DNS information for all members is lost
7. debugging and detection when DNS information for one member is lost
8. Removal of a member from the etcd cluster
9. Recovery of a member with a bad data-dir.
10. Addition of a new member when there is significant etcd data.
11. Upgrade, downgrade, re-upgrade

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+*  Recover from expired certificates (not considered for 4.4)
+
+## Proposal
+
+## User Stories [optional]
+
+## Implementation Plan
+
+1. Make changes to the scripts as needed
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: links to specific scripts please

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+## Motivation
+
+The immediate need for enhancement of disaster recovery scripts is to allow basic backup and restore operations to work, which are currently failing the automated tests (e2edisruptive tests). 
+
+In addition, the disaster recovery should allow the recovery of a single failed node, scaling up and scaling down (node drain).
+
+## Goals
+
+*  Backing up etcd data, including etcd static-pod-resources
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: numbers please.

Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: do we provide backup scripts today?

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+## User Stories [optional]
+
+## Implementation Plan
+
+1. Make changes to the scripts as needed
+2. Modify CEO to provide the information needed for the DR scripts to work.
+
+## Implementation Details/Notes/Constraints
+
+*  Fix the version mismatch bug
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: link please. Is this a design problem?

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+*  Change the name of the spec from `etcd-member.yaml` to `etcd-pod.yaml`
+*  Change the backup scripts to include the static pod resources of etcd
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: Can you write out the logical steps required for our cases?  Most people reading aren't familiar with the vagaries of the current script and it such text age well.

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+Current disaster recovery scripts were created before the introduction of `cluster-etcd-operator` (CEO). Cluster-etcd-operator (CEO) is an operator that handles the scaling of etcd during cluster bootstrap and regular operation and provisioning etcd dependencies such as TLS certificates. This enhancement allows modifications to the DR scripts such that disaster recovery makes use of the presence of CEO.
+
+
+## Motivation
+
+The immediate need for enhancement of disaster recovery scripts is to allow basic backup and restore operations to work, which are currently failing the automated tests (e2edisruptive tests). 
+
+In addition, the disaster recovery should allow the recovery of a single failed node, scaling up and scaling down (node drain).
+
+## Goals
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: missing goal: remove responsibility for these files from the MCO.

File: enhancements/etcd/disaster-recovery-with-ceo.md (hunk 2 above)
Comment (retroflexer) [reply enhancements/etcd/disaster-recovery-with-ceo.md]: yes.

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+*  Recover from expired certificates (not considered for 4.4)
+
+## Currently Supported Functionality 
+
+Currently the scripts available support the following functionality:
+### Backup ([usr-local-bin-etcd-snapshot-backup-sh.yaml](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-etcd-snapshot-backup-sh.yaml))
+Takes a snapshot of cluster’s etcd data along with static-pod-resources at the time of the backup. 
+
+### Restore ([usr-local-bin-etcd-snapshot-restore-sh.yaml](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-etcd-snapshot-restore-sh.yaml))
+Restores the etcd data from a backup snapshot. It also restores the static pod resources while deleting all the newer revisions.
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: how does it do the restoration.  Links don't age well with intent.  Write out in 10 steps what the script does.

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+## Non-Goals
+
+*  Recover from expired certificates (not considered for 4.4)
+
+## Currently Supported Functionality 
+
+Currently the scripts available support the following functionality:
+### Backup ([usr-local-bin-etcd-snapshot-backup-sh.yaml](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-etcd-snapshot-backup-sh.yaml))
+Takes a snapshot of cluster’s etcd data along with static-pod-resources at the time of the backup.
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: how does it do the snapshot. Links don't age well with intent. Write out in 10 steps what the script does.

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+### Restore ([usr-local-bin-etcd-snapshot-restore-sh.yaml](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-etcd-snapshot-restore-sh.yaml))
+Restores the etcd data from a backup snapshot. It also restores the static pod resources while deleting all the newer revisions.
+
+### Remove an etcd member ([usr-local-bin-etcd-member-remove-sh.yaml](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-etcd-member-remove-sh.yaml))
+This script removes a member from the etcd cluster membership. This is useful when trying to replace a failed master host.
+
+### Add an etcd member ([usr-local-bin-etcd-member-add-sh.yaml](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-etcd-member-add-sh.yaml))
+This script adds a member to the etcd cluster membership. This script assumes the certificates are valid.
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: Why do we have this script?  Why wouldn't a cluster-admin just run cluster-add.  Do we even plan to keep this?

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+ 
+### Replacing a single failed member 
+If some etcd members fail, but you still have a quorum of etcd members, you can use the remaining etcd members and the data that they contain to add more etcd members without etcd or cluster downtime. Just running the remove script, and the add script will readd the member to the cluster.
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: line breaks every sentence or 120ish characters, whichever is shorter.

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: this suggests that add and remove scripts are local to machines in some way?  I think you should sketch out the 10 steps or less to make this work on 4.4.  Locally adding a member won't play well with the clustermembercontroller

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+### Recover an etcd member ([usr-local-bin-etcd-member-recover-sh.yaml](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-etcd-member-recover-sh.yaml))
+This script helps recover from a complete loss of a master host. This includes situations where a majority of master hosts have been lost, leading to etcd quorum loss and the cluster going offline. This procedure assumes that you have at least one healthy master host. It restores etcd data from backup to gain etcd quorum on the active master host. The steps to recover include setting up an etcd-signer to obtain certificates for new master hosts and growing them to full membership.
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: consider whether this is truly needed.  If the host is entirely gone, can we just rsh to an etcd pod, do a remove, then add the new member?

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+### Recover from expired certificates ([recover-kubeconfig.sh](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-openshift-kubeconfig-gen.yaml))
+This procedure enables to recover from a situation where the control plane certificates have expired.
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: which control plane certificates do you think can expire?  I'm pretty sure we mint for 10 years.

File: enhancements/etcd/disaster-recovery-with-ceo.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+Currently the scripts available support the following functionality:
+### Backup ([usr-local-bin-etcd-snapshot-backup-sh.yaml](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-etcd-snapshot-backup-sh.yaml))
+Takes a snapshot of cluster’s etcd data along with static-pod-resources at the time of the backup. 
+
+### Restore ([usr-local-bin-etcd-snapshot-restore-sh.yaml](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-etcd-snapshot-restore-sh.yaml))
+Restores the etcd data from a backup snapshot. It also restores the static pod resources while deleting all the newer revisions.
+
+### Remove an etcd member ([usr-local-bin-etcd-member-remove-sh.yaml](https://github.com/openshift/machine-config-operator/blob/3e10747951dd1c3c7a9f131d4c5af7805f19a164/templates/master/00-master/_base/files/usr-local-bin-etcd-member-remove-sh.yaml))
+This script removes a member from the etcd cluster membership. This is useful when trying to replace a failed master host.
```
Comment (deads2k) [inline enhancements/etcd/disaster-recovery-with-ceo.md]: why is this better than `oc -n openshift-etcd rsh pod/working-pod -- ectdctl member remove bad-member`?

**Comment by deads2k** [comment]
assuming this is the actual state

/lgtm

Another PR should be open to flesh out my question for 4.4

**Comment by retroflexer** [comment]
> Another PR should be open to flesh out my question for 4.4
Yes. @alaypatel07  is working on the other PR. Thanks.

**Pull request merged**